File: backend/src/presentation/schemas/stt.py
"""STT API Schemas.

Feature: 003-stt-testing-module
Schemas for STT endpoints matching frontend expectations.
"""

from typing import Literal

from pydantic import Field

from src.presentation.schemas.common import BaseSchema

# --- Provider Schemas ---


class STTProviderResponse(BaseSchema):
    """STT provider information matching frontend STTProvider type."""

    name: str = Field(..., description="Provider identifier (azure, gcp, whisper)")
    display_name: str = Field(..., description="Human-readable name")
    supports_streaming: bool = Field(..., description="Whether provider supports streaming")
    supports_child_mode: bool = Field(..., description="Whether provider supports child mode")
    max_duration_sec: int = Field(..., description="Maximum audio duration in seconds")
    max_file_size_mb: int = Field(..., description="Maximum file size in MB")
    supported_formats: list[str] = Field(..., description="Supported audio formats")
    supported_languages: list[str] = Field(..., description="Supported language codes")
    has_credentials: bool = Field(
        default=False, description="Whether user has credentials configured"
    )
    is_valid: bool = Field(default=False, description="Whether configured credentials are valid")


class STTProvidersListResponse(BaseSchema):
    """Response for GET /stt/providers."""

    providers: list[STTProviderResponse]


# --- Word Timing Schemas ---


class WordTimingResponse(BaseSchema):
    """Word timing information matching frontend WordTiming type."""

    word: str = Field(..., description="The transcribed word")
    start_ms: int = Field(..., description="Start time in milliseconds")
    end_ms: int = Field(..., description="End time in milliseconds")
    confidence: float | None = Field(default=None, description="Confidence score 0-1")


# --- Transcription Schemas ---


class STTTranscribeRequest(BaseSchema):
    """Request for STT transcription."""

    provider: str = Field(..., description="STT provider name (gcp, azure, whisper)")
    language: str = Field(default="zh-TW", description="Language code")
    child_mode: bool = Field(default=False, description="Enable child speech mode")
    ground_truth: str | None = Field(default=None, description="Ground truth for WER calculation")
    save_to_history: bool = Field(default=True, description="Save to test history")


class WERAnalysisRequest(BaseSchema):
    """Request for WER/CER analysis."""

    result_id: str = Field(..., description="Transcription result ID")
    ground_truth: str = Field(..., description="Ground truth text")


class WERAnalysisResponse(BaseSchema):
    """WER/CER analysis result matching frontend WERAnalysis type."""

    error_rate: float = Field(..., description="Error rate (0-1)")
    error_type: Literal["WER", "CER"] = Field(..., description="Type of error rate")
    insertions: int = Field(..., description="Number of inserted words/chars")
    deletions: int = Field(..., description="Number of deleted words/chars")
    substitutions: int = Field(..., description="Number of substituted words/chars")
    total_reference: int = Field(..., description="Total words/chars in reference")


class STTTranscribeResponse(BaseSchema):
    """Response from STT transcription matching frontend TranscriptionResponse."""

    id: str | None = Field(default=None, description="Transcription result ID")
    transcript: str = Field(..., description="Transcribed text")
    provider: str = Field(..., description="Provider used")
    language: str = Field(..., description="Language detected/used")
    latency_ms: int = Field(..., description="Processing latency in milliseconds")
    confidence: float = Field(default=0.0, description="Transcription confidence 0-1")
    words: list[WordTimingResponse] | None = Field(
        default=None, description="Word-level timing information"
    )
    audio_duration_ms: int | None = Field(default=None, description="Audio duration in ms")
    wer_analysis: WERAnalysisResponse | None = Field(
        default=None, description="WER/CER analysis if ground truth provided"
    )
    created_at: str | None = Field(default=None, description="ISO timestamp")

    # Legacy fields for backwards compatibility
    word_timings: list[WordTimingResponse] | None = Field(
        default=None, description="Deprecated: Use 'words' instead"
    )
    wer: float | None = Field(default=None, description="Deprecated: Use wer_analysis")
    cer: float | None = Field(default=None, description="Deprecated: Use wer_analysis")
    record_id: str | None = Field(default=None, description="Deprecated: Use 'id'")


class TranscriptionSummary(BaseSchema):
    """Summary of a transcription record."""

    id: str
    provider: str
    language: str
    transcript_preview: str
    duration_ms: int | None
    confidence: float
    has_ground_truth: bool
    error_rate: float | None
    created_at: str


class TranscriptionHistoryPage(BaseSchema):
    """Paginated transcription history."""

    items: list[TranscriptionSummary]
    total: int
    page: int
    page_size: int
    total_pages: int


class AudioFileResponse(BaseSchema):
    """Audio file details."""

    id: str
    filename: str
    duration_ms: int
    format: str
    download_url: str | None = None


class TranscriptionDetail(STTTranscribeResponse):
    """Detailed transcription record."""

    audio_file: AudioFileResponse
    ground_truth: str | None
    child_mode: bool


class ComparisonResponse(BaseSchema):
    """Comparison results for multiple providers."""

    audio_file_id: str
    results: list[STTTranscribeResponse]
    ground_truth: str | None
    comparison_table: list[dict]

File: frontend/src/types/stt.ts
/**
 * STT (Speech-to-Text) TypeScript types for the STT Testing Module.
 * Feature: 003-stt-testing-module
 */

// Provider types
export type STTProviderName = 'azure' | 'gcp' | 'whisper'

export interface STTProvider {
  name: STTProviderName
  display_name: string
  supports_streaming: boolean
  supports_child_mode: boolean
  max_duration_sec: number
  max_file_size_mb: number
  supported_formats: AudioFormat[]
  supported_languages: string[]
  /** Whether the user has configured credentials for this provider */
  has_credentials?: boolean
  /** Whether the configured credentials are valid */
  is_valid?: boolean
}

// Audio types
export type AudioFormat = 'mp3' | 'wav' | 'm4a' | 'flac' | 'webm' | 'ogg'

export interface AudioFile {
  id: string
  filename: string
  format: AudioFormat
  duration_ms: number
  sample_rate: number
  file_size_bytes: number
  source: 'upload' | 'recording'
  created_at: string
}

// Request types
export interface STTTranscribeRequest {
  provider: STTProviderName
  language: string
  child_mode?: boolean
  ground_truth?: string
}

// Response types
export interface WordTiming {
  word: string
  start_ms: number
  end_ms: number
  confidence?: number
}

export interface WERAnalysis {
  error_rate: number
  error_type: 'WER' | 'CER'
  insertions: number
  deletions: number
  substitutions: number
  total_reference: number
  alignment?: AlignmentItem[]
}

export interface AlignmentItem {
  ref: string | null
  hyp: string | null
  op: 'match' | 'substitute' | 'insert' | 'delete'
}

export interface TranscriptionResponse {
  id: string
  provider: STTProviderName
  transcript: string
  confidence: number | null
  latency_ms: number
  language: string
  words?: WordTiming[]
  wer_analysis?: WERAnalysis
  audio_duration_ms?: number
  created_at: string
}

// History types
export interface TranscriptionSummary {
  id: string
  provider: STTProviderName
  language: string
  transcript_preview: string
  duration_ms?: number
  confidence: number | null
  has_ground_truth: boolean
  error_rate?: number
  created_at: string
}

export interface TranscriptionHistoryPage {
  items: TranscriptionSummary[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

export interface TranscriptionDetail extends TranscriptionResponse {
  audio_file?: {
    id: string
    filename: string
    duration_ms: number
    format: AudioFormat
    download_url: string
  }
  ground_truth?: string
  child_mode: boolean
}

// Comparison types
export interface ComparisonResult {
  provider: STTProviderName
  transcript: string
  confidence: number | null
  latency_ms: number
  error_rate?: number
  error_type?: 'WER' | 'CER'
}

export interface ComparisonResponse {
  audio_file_id: string
  results: TranscriptionResponse[]
  ground_truth?: string
  comparison_table: ComparisonResult[]
}

// UI State types
export interface STTTestState {
  // Audio input
  audioFile: File | null
  audioBlob: Blob | null
  audioUrl: string | null
  isRecording: boolean

  // Provider selection
  selectedProvider: STTProviderName
  availableProviders: STTProvider[]

  // Options
  language: string
  childMode: boolean
  groundTruth: string

  // Results
  isTranscribing: boolean
  transcriptionResult: TranscriptionResponse | null
  werAnalysis: WERAnalysis | null
  error: string | null

  // Comparison
  comparisonProviders: STTProviderName[]
  comparisonResults: ComparisonResponse | null
  isComparing: boolean
}

// Error types
export interface STTError {
  error: string
  detail: string
  code: 'INVALID_FORMAT' | 'FILE_TOO_LARGE' | 'PROVIDER_ERROR' | 'QUOTA_EXCEEDED'
}

// Supported languages
export const SUPPORTED_LANGUAGES = [
  { code: 'zh-TW', name: '繁體中文 (台灣)', errorType: 'CER' },
  { code: 'zh-CN', name: '簡體中文', errorType: 'CER' },
  { code: 'en-US', name: 'English (US)', errorType: 'WER' },
  { code: 'ja-JP', name: '日本語', errorType: 'CER' },
  { code: 'ko-KR', name: '한국어', errorType: 'CER' },
] as const

export type SupportedLanguage = (typeof SUPPORTED_LANGUAGES)[number]['code']
